plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
+        $visitFieldsToSelect = array_merge(array('idvisit', 'location_ip'), array_keys(self::$logVisitFieldsToUpdate));
+
+        $lastId = 0;
+        do {
+            $logs = $this->dao->getVisitsWithDatesLimit($from, $to, $visitFieldsToSelect, $lastId, $segmentLimit);
+            if (!empty($logs)) {
+                $lastId = $logs[count($logs) - 1]['idvisit'];
+
+                $this->reattributeVisitLogs($output, $logs);
+            }
+        } while (count($logs) == $segmentLimit);
+    }
+
+    protected function reattributeVisitLogs(OutputInterface $output, $logRows)
+    {
+        foreach ($logRows as $row) {
+            if (!$this->isRowComplete($output, $row)) {
+                continue;
+            }
+
+            $location = $this->getVisitLocation($row);
+            $valuesToUpdate = $this->getValuesToUpdate($row, $location);
+
+            $this->onVisitProcessed($output);
+
+            $idVisit = $row['idvisit'];
+
+            if (empty($valuesToUpdate)) {
+                $this->writeIfVerbose($output, 'Nothing to update for idvisit = ' . $idVisit . '. Existing location info is same as geolocated.');
+            } else {
+                $this->writeIfVerbose($output, 'Updating visit with idvisit = ' . $idVisit . '.');
+
+                $this->dao->updateVisits($valuesToUpdate, $idVisit);
+                $this->dao->updateConversions($valuesToUpdate, $idVisit);
+            }
+        }
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return int
+     */
+    protected function getPercentStep(InputInterface $input)
+    {
+        $percentStep = $input->getArgument(self::PERCENT_STEP_ARGUMENT);
+
+        if (!is_numeric($percentStep)) {
+            return self::PERCENT_STEP_ARGUMENT_DEFAULT;
+        }
+
+        // Percent step should be between maximum percent value and minimum percent value (1-100)
+        if ($percentStep > 99 || $percentStep < 1) {
+            return 100;
+        }
+
+        return $percentStep;
+    }
+
+    /**
+     * Returns location log values that are different than the values currently in a log row.
+     *
+     * @param array $row The visit row.
+     * @param array $location The location information.
+     * @return array The location properties to update.
+     */
+    protected function getValuesToUpdate(array $row, $location)
```

Comment:
I don't think it would suggest that all callers should use this method. That's just a matter of code organization etc (eg where the class is placed and how it is named). The alternative is, that this code/method is in a command class where it doesn't really belong to (It's basically a Controller and shouldn't have any logic etc. It should only get some input params, trigger some action/business logic and care about output). Also in the end you **probably** want to have it testable so a solution would be to extract it. The advantage outweighs the disadvantage I think but it's surely up to you. I do not really care whether it uses an iterator. If an iterator is the best solution to solve it, use it ;)